sdk/go/pulumi-language-go/main.go

```diff
@@ -352,7 +352,25 @@ func (host *goLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest)
 			// If the program ran, but exited with a non-zero error code.  This will happen often, since user
 			// errors will trigger this.  So, the error message should look as nice as possible.
 			if status, stok := exiterr.Sys().(syscall.WaitStatus); stok {
-				err = errors.Errorf("program exited with non-zero exit code: %d", status.ExitStatus())
+				switch status.ExitStatus() {
+				case 0:
+					// This really shouldn't happen, but if it does, we don't want to render "non-zero exit code"
+					err = errors.Wrapf(exiterr, "Program exited unexpectedly")
+				case 1:
+					// runtime error(panic or error return value)
+					return &pulumirpc.RunResponse{Error: "", Bail: true}, nil
+				case 2:
```

Comment:
This is weird. It seems that `go run main.go` hijacks the exit code. So it would return `2` if go compilation failed, and `1` if program failed. The real exit code it would print instead of returning.

But this code should not assume it's running `go run`. It may be running precompiled binary. In that case exit code 2 is not a compilation error. Can we handle that?